src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/ZLibStreamPool.cs

```diff
@@ -0,0 +1,287 @@
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    internal sealed class ZLibStreamPool
+    {
+        private static readonly ZLibStreamPool?[] s_pools
+            = new ZLibStreamPool[WebSocketValidate.MaxDeflateWindowBits - WebSocketValidate.MinDeflateWindowBits + 1];
+
+        /// <summary>
+        /// The default amount of time after which a cached item will be removed.
+        /// </summary>
+        private const int DefaultTimeoutMilliseconds = 60_000;
+
+        private readonly int _windowBits;
+        private readonly List<CacheItem> _inflaters = new();
+        private readonly List<CacheItem> _deflaters = new();
+        private readonly Timer _cleaningTimer;
+
+        /// <summary>
+        /// The amount of time after which a cached item will be removed.
+        /// </summary>
+        private readonly int _timeoutMilliseconds;
+
+        /// <summary>
+        /// The number of cached inflaters and deflaters.
+        /// </summary>
+        private int _activeCount;
+
+        private ZLibStreamPool(int windowBits, int timeoutMilliseconds)
+        {
+            // Use negative window bits to for raw deflate data
+            _windowBits = -windowBits;
+            _timeoutMilliseconds = timeoutMilliseconds;
+
+            bool restoreFlow = false;
+            try
+            {
+                if (!ExecutionContext.IsFlowSuppressed())
+                {
+                    ExecutionContext.SuppressFlow();
+                    restoreFlow = true;
+                }
+
+                // There is no need to use weak references here, because these pools are kept
+                // for the entire lifetime of the application. Also we reset the timer on each tick,
+                // which prevents the object being rooted forever.
+                _cleaningTimer = new Timer(x => ((ZLibStreamPool)x!).RemoveStaleItems(),
+                    state: this, Timeout.Infinite, Timeout.Infinite);
+            }
+            finally
+            {
+                if (restoreFlow)
+                {
+                    ExecutionContext.RestoreFlow();
+                }
+            }
+        }
+
+        public static ZLibStreamPool GetOrCreate(int windowBits)
+        {
+            Debug.Assert(windowBits >= WebSocketValidate.MinDeflateWindowBits
+                      && windowBits <= WebSocketValidate.MaxDeflateWindowBits);
+
+            int index = windowBits - WebSocketValidate.MinDeflateWindowBits;
+            ref ZLibStreamPool? pool = ref s_pools[index];
+
+            return Volatile.Read(ref pool) ?? EnsureInitialized(windowBits, ref pool);
+
+            static ZLibStreamPool EnsureInitialized(int windowBits, ref ZLibStreamPool? target)
+            {
+                Interlocked.CompareExchange(ref target, new ZLibStreamPool(windowBits, DefaultTimeoutMilliseconds), null);
```

Comment:
what if another thread was lucky and initialized `target` first?

```suggestion
                var newPool = new ZLibStreamPool(windowBits, DefaultTimeoutMilliseconds);
                if (Interlocked.CompareExchange(ref target, newPool, null) is not null)
                {
                    newPool._cleaningTimer.Dispose();
                }
```